docs/syntax/Cadence.syntax

```diff
@@ -0,0 +1,1482 @@
+    ;
+
+execute:
+        identifier block
+    ;
+
+importDeclaration:
+        importKeyword *1importDeclarationFromPhrase importDeclarationSource
+    ;
+
+importDeclarationFromPhrase:
+        1*identifier( "," ) fromKeyword
+    ;
+
+importDeclarationSource:
+        stringLiteral
+    |   hexadecimalLiteral
+    |   identifier
+    ;
+
+access:
+        accessPriv
+    |   accessPub
+    |   accessSpecified
+    |   accessNotSpecified
+    ;
+
+accessNotSpecified:
+    ;
+
+accessPriv:
+        privKeyword
+    ;
+
+accessPub:
+        pubKeyword *1accessPubSet
+    ;
+
+accessPubSet:
+        "(" setKeyword ")"
+    ;
+
+accessSpecified:
+        accessKeyword "(" accessAccessValue ")"
+    ;
+
+accessAccessValue:
+        selfKeyword
+    |   contractKeyword
+    |   accountKeyword
+    |   allKeyword
+    |   accessAccessValueEntitlement
+    |   accessAccessValueEntitlementMapping
+    ;
+
+accessAccessValueEntitlement:
+        accessAccessValueEntitlementConjunction
+    |   accessAccessValueEntitlementDisjunction
+    ;
+
+accessAccessValueEntitlementConjunction:
+        1*fullType( "," )
+    ;
+
+accessAccessValueEntitlementDisjunction:
+        1*fullType( "|" )
+    ;
+
+accessAccessValueEntitlementMapping:
+        mappingKeyword fullType
+    ;
+
+conformances:
+        conformancesTypes
+    |   conformancesEmpty
+    ;
+
+conformancesTypes:
+        ":" 1*nominalType( "," )
+    ;
+
+conformancesEmpty:
+    ;
+
+field:
+        letField
+    |   varField
+    |   unspecifiedField
+    ;
+
+letField:
+        access letKeyword nonReservedIdentifier ":" typeAnnotation
+    ;
+
+varField:
+        access varKeyword nonReservedIdentifier ":" typeAnnotation
+    ;
+
+unspecifiedField:
+        access nonReservedIdentifier ":" typeAnnotation
+    ;
+
+fields:
+        *fieldsEntry
+    ;
+
+fieldsEntry:
+        field *1semicolon
+    ;
+
+structCompositeDeclaration:
+        access structKeyword nonReservedIdentifier conformances "{" membersAndNestedDeclarations "}"
+    ;
+
+resourceCompositeDeclaration:
+        access resourceKeyword nonReservedIdentifier conformances "{" membersAndNestedDeclarations "}"
+    ;
+
+contractCompositeDeclaration:
+        access contractKeyword nonReservedIdentifier conformances "{" membersAndNestedDeclarations "}"
+    ;
+
+enumCompositeDeclaration:
+        access enumKeyword nonReservedIdentifier conformances "{" *semicolonEnumerationDeclarationCase "}"
+    ;
+
+structInterfaceDeclaration:
+        access structKeyword interfaceKeyword nonReservedIdentifier *1interfaceInheritance "{" membersAndNestedDeclarations "}"
+    ;
+
+resourceInterfaceDeclaration:
+        access resourceKeyword interfaceKeyword nonReservedIdentifier *1interfaceInheritance "{" membersAndNestedDeclarations "}"
+    ;
+
+contractInterfaceDeclaration:
+        access contractKeyword interfaceKeyword nonReservedIdentifier *1interfaceInheritance "{" membersAndNestedDeclarations "}"
+    ;
+
+enumInterfaceDeclaration:
```

Comment:
We might want to indicate that enum interfaces are illegal, but this rule exists to produce better error messages. We might in general make this syntax definition as strict as possible, so maybe just remove the rule

Reply:
I was thinking we should have two versions of the syntax. A strict grammar that is allowed by the compiler, and a more general, internal syntax that is used to produce useful error messages.

Reply:
That's a good idea. To begin with we might at least indicate which variant of the syntax this is